src/Microsoft.AspNet.SignalR.Client/Hubs/HubProxy.cs

```diff
@@ -166,12 +166,20 @@ public Task Invoke<T>(string method, Action<T> onProgress, params object[] args)
                 CallbackId = callbackId
             };
 
-            if (_state.Count != 0)
+            string value = null;
+
+            lock (_state)
             {
-                hubData.State = _state;
+                if (_state.Count != 0)
+                {
+                    hubData.State = _state;
+
+                    // Only keep the _state lock during JsonSerializeObject if hubData.State is set to _state.
+                    value = _connection.JsonSerializeObject(hubData);
+                }
             }
 
-            var value = _connection.JsonSerializeObject(hubData);
+            value = value ?? _connection.JsonSerializeObject(hubData);
```

Comment:
would there be any harm in putting all `_connection.JsonSerializeObject()` calls in the lock?
```csharp
lock (_state)
{
    if (_state.Count != 0)
    {
        hubData.State = _state;
    }
    value = _connection.JsonSerializeObject(hubData);
}
```

Reply:
I don't think the lock will actually be contested in the vast majority of apps, but theoretically Invoke can be called from multiple threads, so I figure why take the risk of acquiring the lock for longer than necessary in the normal case when the state is empty.

Reply:
After another look it doesn't make sense to do the other serialize inside the lock, I was thinking the `hubData.State` property could be set outside the lock, but it's a local.